### PR TITLE
fix(ingest/s3): Path spec aware folder traversal

### PR DESCRIPTION
**Pull request by treff7es**
Folder traversal needs to be aware of the path_spec and not go into folders which should not match with the path_spec.
The logic heavily relies on that nature of path spec which mandates to specify each folder level. 

This pr depends on #8089 

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/tests/integration/s3/test_data/local_system/folder_a/folder_aa/folder_aaa/pokemon_abilities_json/year=2022/month=jan/_temporary/dummy.json
Comment (mayurinehate) [inline metadata-ingestion/tests/integration/s3/test_data/local_system/folder_a/folder_aa/folder_aaa/pokemon_abilities_json/year=2022/month=jan/_temporary/dummy.json:1]: Is this file added intentionally ?

File: metadata-ingestion/src/datahub/ingestion/source/data_lake_common/data_lake_utils.py [hunk 1]
```diff
@@ -74,11 +79,28 @@ def gen_bucket_key(self, name):
...
+        if is_s3_uri(path):
+            protocol = get_s3_prefix(path)
+        elif is_gcs_uri(path):
+            protocol = get_gcs_prefix(path)
+
+        if protocol:
+            return protocol
+        else:
+            raise ValueError(
+                f"Unable to get protocol or invalid protocol form path: {path}"
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/data_lake_common/data_lake_utils.py:94]: typo form -> from

File: metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py [hunk 2]
```diff
@@ -81,6 +81,30 @@ def allowed(self, path: str) -> bool:
...
+        path_slash = path.count("/")
+        glob_slash = self.glob_include.count("/")
+        if path_slash > glob_slash:
+            return False
+        slash_to_remove_from_glob = glob_slash - path_slash
+
+        glob_include = self.glob_include.rsplit("/", 1)[0]
+
+        for i in range(slash_to_remove_from_glob):
+            glob_include = glob_include.rsplit("/", 1)[0]
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py:94]: Could we get this via `glob_include.rsplit("/", i)[0]`, so I guess `glob_include.rsplit("/", slash_to_remove_from_glob + 1)[0]` in total? If that's the first time it's defined

**Review by asikowitz — APPROVED** [review]
A couple minor requests but don't think this needs a rereview.

File: metadata-ingestion/tests/integration/s3/test_data/local_system/folder_a/folder_aa/folder_aaa/pokemon_abilities_json/year=2022/month=jan/_temporary/dummy.json
Comment (treff7es) [reply metadata-ingestion/tests/integration/s3/test_data/local_system/folder_a/folder_aa/folder_aaa/pokemon_abilities_json/year=2022/month=jan/_temporary/dummy.json:1]: Yes, without the new logic, this file would have been picked up.

Comment (mayurinehate) [reply metadata-ingestion/tests/integration/s3/test_data/local_system/folder_a/folder_aa/folder_aaa/pokemon_abilities_json/year=2022/month=jan/_temporary/dummy.json:1]: okay, nice.

File: metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py [hunk 3]
```diff
@@ -81,6 +81,30 @@ def allowed(self, path: str) -> bool:
...
+
```
[the remark this answers is not in the document]
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py:95]: I added a comment and restructured the code to make it more clear.

File: metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py (hunk 2 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/data_lake_common/path_spec.py:98]: Thanks, I made it more clear in the code.

**Pull request merged**